src/prognostic_equations/implicit/autodiff_utils.jl

```diff
@@ -0,0 +1,92 @@
+    hs = Spaces.eachslabindex(horz_space)
+    return horz_space isa Spaces.SpectralElementSpace1D ?
+           Iterators.product(qs, hs) : Iterators.product(qs, qs, hs)
+end
+
+# An iterator over the FieldNames of all scalar fields in a FieldVector.
+scalar_field_names(field_vector) =
+    MatrixFields.filtered_names(field_vector) do x
+        x isa Fields.Field && eltype(x) == eltype(field_vector)
+    end
+
+# An iterator with pairs of the form
+# (scalar_level_index, (scalar_index, level_index)), where scalar_index is an
+# index into scalar_field_names(field_vector), level_index is a vertical index
+# into the scalar field with this name, and scalar_level_index is a unique
+# linear index for each combination of scalar field name and vertical index.
+function scalar_level_index_pairs(field_vector)
+    scalar_names = scalar_field_names(field_vector)
+    scalar_index_and_level_pairs =
+        unrolled_map(enumerate(scalar_names)) do (scalar_index, name)
+            field = MatrixFields.get_field(field_vector, name)
+            is_one_level = field isa Fields.SpectralElementField
+            Iterators.map(
+                level_index -> (scalar_index, level_index),
+                Base.OneTo(is_one_level ? 1 : Spaces.nlevels(axes(field))),
+            )
+        end
+    return enumerate(Iterators.flatten(scalar_index_and_level_pairs))
+end
+
+# A view of the data for one point in a field, which can be used like a Ref.
+Base.@propagate_inbounds function point(field, level_index, column_index...)
+    column_field = Fields.column(field, column_index...)
+    column_field isa Fields.PointField && return column_field
+    level_index_offset = Operators.left_idx(axes(column_field)) - 1
+    return Fields.level(column_field, level_index + level_index_offset)
+end
+
+# TODO: This needs to be moved into ClimaCore to avoid breaking precompilation.
+# function ClimaComms.context(x::Fields.FieldVector)
+#     values = Fields._values(x)
+#     isempty(values) && error("Empty FieldVector has no device or context")
+#     index = unrolled_findfirst(Base.Fix2(!isa, Fields.PointField), values)
+#     return ClimaComms.context(values[isnothing(index) ? 1 : index])
+# end
+
+##################################################
+## Nonstandard rules for automatic differentiation
+##################################################
+
+# Every nonstandard rule should be defined using a method that specializes on
+# the tag Jacobian. Not specializing on any tag overwrites the generic method
+# for Dual in ForwardDiff and breaks precompilation, while specializing on the
+# default tag Nothing causes the type piracy test to fail.
+
+# Set the derivative of sqrt(x) to iszero(x) ? zero(x) : inv(2 * sqrt(x)) in
+# order to properly handle derivatives of x * sqrt(x). Without this change, the
+# derivative of the turbulent kinetic energy dissipation tendency, which is a
+# linear function of ᶜtke⁰ * sqrt(ᶜtke⁰), is NaN whenever tke⁰ = 0 at any point.
```

Comment:
If we express the tke dissipation tendency as tke^1.5 would it solve the problem without overriding sqrt derivative?